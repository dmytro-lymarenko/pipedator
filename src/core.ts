export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
