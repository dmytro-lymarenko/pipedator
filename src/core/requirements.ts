export interface BaseRequirement {
	value: any;
	path: string[];
	message: string;
}

export interface RequirementFactory<R extends Requirement = Requirement> {
	(path: string[], value: any): R;
}

export type Requirement = SingleRequirement | DependenceRequirement | AlternativeRequirement | ConcatenationRequirement;

export interface SingleRequirement extends BaseRequirement {
	type: 'single';
}

export interface DependenceRequirement extends BaseRequirement {
	type: 'dependence';
	requirement: Requirement;
}

export interface AlternativeRequirement extends BaseRequirement {
	type: 'alternative';
	requirements: Requirement[];
}

export interface ConcatenationRequirement extends BaseRequirement {
	type: 'concatenation';
	requirements: Requirement[];
}

export function isRequirement(requirement: Requirement | null): requirement is Requirement {
	return requirement !== null;
}

// Factories

export function singleRequirementFactory(message: string): RequirementFactory<SingleRequirement> {
	return (path, value) => ({
		path,
		value,
		message,
		type: 'single',
	});
}

export function dependenceRequirementFactory(
	message: string,
	requirement: Requirement
): RequirementFactory<DependenceRequirement> {
	return (path, value) => ({
		path,
		value,
		requirement,
		message,
		type: 'dependence',
	});
}

export function alternativeRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<AlternativeRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'alternative',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a || (b || (c && d)) || e) === (a || b || (c && d) || e)
				requirement.type === 'alternative' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

export function concatenationRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<ConcatenationRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'concatenation',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a && (b && (c || d)) && e) === (a && b && (c || d) && e)
				requirement.type === 'concatenation' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

// examples:
// const numberRequirementFactory = singleRequirementFactory('Value should be a number');

// function equalToRequirementFactory(validValue: any): RequirementFactory<SingleRequirement> {
// 	return singleRequirementFactory(`Value should equal to ${validValue}`);
// }

// function optionalRequirementFactory(requirement: Requirement): RequirementFactory<AlternativeRequirement> {
// 	const undefinedFactory = equalToRequirementFactory(undefined);

// 	return (path, value) => alternativeRequirementFactory('', [undefinedFactory(path, value), requirement])(path, value);
// }

// const optionalNumberRequirementFactory: RequirementFactory<AlternativeRequirement> = (path, value) =>
// 	optionalRequirementFactory(numberRequirementFactory(path, value))(path, value);

// function someRequirementFactory(requirement: Requirement): RequirementFactory<DependenceRequirement> {
// 	return dependenceRequirementFactory('Some item should follow the requirement', requirement);
// }
