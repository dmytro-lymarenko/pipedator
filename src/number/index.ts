import { test } from '../test';

export function number<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'number', message || 'Value should be a number');
}
