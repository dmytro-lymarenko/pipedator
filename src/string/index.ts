import { test } from '../test';

export function string<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'string', message || 'Value should be a string');
}
