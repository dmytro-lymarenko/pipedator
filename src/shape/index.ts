/**
 * @module validators
 */

import { Validator } from '../core';
import { abstractShape } from '../abstractShape';

export function shape(shape: { [key: string]: Validator }, options?: { onlyFirstError?: boolean }, message?: string) {
	const keys = Object.keys(shape);

	return abstractShape(keys, key => shape[key], options, message);
}
