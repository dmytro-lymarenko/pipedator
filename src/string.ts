import { createValidator } from './core';

export function string() {
	return createValidator({
		message: 'Value should be a string',
		validate: (value, ctx) => (typeof value === 'string' ? null : { value, message: ctx.message, path: ctx.path }),
	});
}

export function empty() {
	return createValidator({
		message: 'Value should be empty',
		validate: (value, ctx) => (value.length === 0 ? null : { value, message: ctx.message, path: ctx.path }),
	});
}
