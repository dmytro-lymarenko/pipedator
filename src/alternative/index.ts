import { createValidator, getFirstRequirements, Validator } from '../core';
import { alternativeRequirementFactory } from '../core/requirements';

export function alternative<ValidValue = any>(validators: Validator[], message?: string) {
	return createValidator<ValidValue>({
		validate: (value, ctx) => {
			if (validators.length === 0) {
				return null;
			}
			// find the first success validator
			const requirements = getFirstRequirements(i => validators[i], () => value, () => ctx, validators.length);

			if (requirements.length === validators.length) {
				// here all validators returned an error
				if (requirements.length === 1) {
					return requirements[0];
				}

				return alternativeRequirementFactory(
					message || `Either ${requirements.map(requirement => requirement.message).join(' OR ')}`,
					requirements
				)(ctx.path, value);
			}

			return null;
		},
	});
}
