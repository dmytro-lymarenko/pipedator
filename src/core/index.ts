import { Requirement } from './requirements';

export interface ValidationError {
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
	message: string;
	value: any; // the value at current path
	errors: ValidationError[];
}

export interface ValidationErrorOptions {
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
	message: string;
	value: any; // the value at current path
	errors?: ValidationError[];
}

export interface GroupedValidationError {
	errors: string[];
	path: string; // an item from the path
	children: GroupedValidationError[] | null;
}

export interface ValidationContext {
	value: any;
	path: string[];
	rootValue: any;
	// generateError: (error: ValidationErrorOptions) => ValidationError;
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | Requirement;
}

export interface Validator<ValidValue = any> {
	validate: (value: any, ctx?: ValidationContext) => null | Requirement;
	isValid: (value: any) => value is ValidValue;
}

export function findFirstRequirement(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): { requirement: Requirement | null; index: number } {
	let i: number = 0;
	let requirement: Requirement | null = null;

	while (i < length && requirement === null) {
		requirement = getValidator(i).validate(getValue(i), getContext(i));
		i = i + 1;
	}

	return { requirement, index: i - 1 };
}

export function getFirstRequirements(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): Requirement[] {
	const requirements: Requirement[] = [];
	let i: number = 0;
	let requirement: Requirement | null = null;

	do {
		requirement = getValidator(i).validate(getValue(i), getContext(i));
		if (requirement) {
			requirements.push(requirement);
		}
		i = i + 1;
	} while (i < length && requirement !== null);

	return requirements;
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator<ValidValue = any>(options: CreateValidatorOptions): Validator<ValidValue> {
	function validate(value: any, ctx?: ValidationContext) {
		const context: ValidationContext = ctx || {
			value,
			path: [],
			rootValue: value,
			// generateError: errorOptions => ({
			// 	path: errorOptions.path,
			// 	message:
			// 		errorOptions.message.startsWith('[') && errorOptions.message.endsWith(']')
			// 			? errorOptions.message
			// 			: `[${errorOptions.message}]`,
			// 	value: errorOptions.value,
			// 	errors: errorOptions.errors || [],
			// }),
		};

		return options.validate(value, context);
	}

	function isValid(value: any): value is ValidValue {
		return validate(value) === null;
	}

	return {
		validate,
		isValid,
	};
}

export function validate(validator: Validator, value: any) {
	const ctx: ValidationContext = {
		value,
		path: [],
		rootValue: value,
		// generateError: errorOptions => ctx.generateError(errorOptions),
	};

	return validator.validate(value, ctx);
}

export function isValidationError(error: ValidationError | null): error is ValidationError {
	return error !== null;
}

export interface ValidationRef {
	(ctx: ValidationContext): any;
}

export function ref(path: string[]): ValidationRef {
	return (ctx: ValidationContext) => path.reduce((v, key) => v && v[key], ctx.rootValue);
}

export function groupErrors(error: Requirement | null): GroupedValidationError | null {
	if (!error) {
		return null;
	}

	const groupedError: GroupedValidationError = {
		path: '',
		errors: [],
		children: null,
	};

	// groupedRequirement is mutable here
	function handle(groupedRequirement: GroupedValidationError, requirement: Requirement) {
		groupedRequirement.errors.push(requirement.message);

		switch (requirement.type) {
			case 'single':
				break;
			case 'dependence':
				handle(groupedRequirement, requirement.requirement);
				break;
			case 'alternative':
			case 'concatenation':
				requirement.requirements.forEach(err => {
					if (err.path.length === requirement.path.length) {
						// paths should be the same here
						// it means this `err` is related to `error`
						handle(groupedRequirement, err);
					} else if (err.path.length > requirement.path.length) {
						const childPath = err.path[err.path.length - 1];

						const childGroupedError = {
							path: childPath,
							errors: [],
							children: null,
						};

						handle(childGroupedError, err);

						groupedRequirement.children = groupedRequirement.children || [];
						groupedRequirement.children.push(childGroupedError);
					}
				});
				break;
		}
	}

	handle(groupedError, error);

	return groupedError;
}
