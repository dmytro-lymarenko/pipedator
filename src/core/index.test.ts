import { createValidator, validate, groupErrors } from './index';
import { shape } from '../shape';
import { string } from '../string';
import { number } from '../number';
import { alternative } from '../alternative';
import { singleRequirementFactory, dependenceRequirementFactory } from '../core/requirements';

describe('createValidator()', () => {
	it('should create validator', () => {
		const validator = createValidator({
			validate: () => null,
		});

		expect(validator).toBeDefined();
		expect(typeof validator.validate).toBe('function');
	});

	it('should call validate function', () => {
		const validateCb = jest.fn(() => null);

		const validator = createValidator({
			validate: validateCb,
		});

		const value = 1;

		const result = validate(validator, value);

		expect(result).toBeNull();
		expect(validateCb.mock.calls.length).toBe(1);
		expect(validateCb.mock.calls[0][0]).toBe(value);
	});

	it('should return what validate function returns', () => {
		expect(
			createValidator({
				validate: (value, ctx) =>
					dependenceRequirementFactory('Test', singleRequirementFactory('Inner requirement')(ctx.path, value))(ctx.path, value),
			}).validate(1)
		).toMatchSnapshot();
	});
});

describe('groupErrors()', () => {
	it('should return null when error is null', () => {
		expect(groupErrors(null)).toBeNull();
	});

	it('should work for single error', () => {
		expect(groupErrors(string().validate(1))).toMatchSnapshot();
	});

	it('should work for single-path errors', () => {
		expect(groupErrors(alternative([string(), number()]).validate({}))).toMatchSnapshot();
	});

	it('should work for complex values', () => {
		expect(
			groupErrors(
				shape({
					email: string(),
					number: number(),
				}).validate(5)
			)
		).toMatchSnapshot();

		expect(
			groupErrors(
				shape({
					email: string(),
					address: shape({
						city: string(),
						house: number(),
					}),
				}).validate(5)
			)
		).toMatchSnapshot();

		expect(
			groupErrors(
				shape({
					email: string(),
					address: shape({
						city: string(),
						house: number(),
					}),
				}).validate({ address: {} })
			)
		).toMatchSnapshot();

		expect(
			groupErrors(
				shape({
					email: string(),
					address: alternative([
						shape({
							city: string(),
							house: alternative([string(), number()]),
						}),
						string(),
					]),
				}).validate({ address: {} })
			)
		).toMatchSnapshot();
	});
});
