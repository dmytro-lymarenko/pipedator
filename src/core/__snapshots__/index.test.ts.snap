// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createValidator() should return what validate function returns 1`] = `
Object {
  "message": "Test",
  "path": Array [],
  "requirement": Object {
    "message": "Inner requirement",
    "path": Array [],
    "type": "single",
    "value": 1,
  },
  "type": "dependence",
  "value": 1,
}
`;

exports[`groupErrors() should work for complex values 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "errors": Array [
        "Value should be a string",
      ],
      "path": "email",
    },
    Object {
      "children": null,
      "errors": Array [
        "Value should be a number",
      ],
      "path": "number",
    },
  ],
  "errors": Array [
    "Value should have a valid shape",
  ],
  "path": "",
}
`;

exports[`groupErrors() should work for complex values 2`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "errors": Array [
        "Value should be a string",
      ],
      "path": "email",
    },
    Object {
      "children": null,
      "errors": Array [
        "Value should be defined",
      ],
      "path": "address",
    },
  ],
  "errors": Array [
    "Value should have a valid shape",
  ],
  "path": "",
}
`;

exports[`groupErrors() should work for complex values 3`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "errors": Array [
        "Value should be a string",
      ],
      "path": "email",
    },
    Object {
      "children": null,
      "errors": Array [
        "Value should be a string",
      ],
      "path": "city",
    },
    Object {
      "children": null,
      "errors": Array [
        "Value should be a number",
      ],
      "path": "house",
    },
  ],
  "errors": Array [
    "Value should have a valid shape",
  ],
  "path": "",
}
`;

exports[`groupErrors() should work for complex values 4`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "errors": Array [
        "Value should be a string",
      ],
      "path": "email",
    },
    Object {
      "children": Array [
        Object {
          "children": null,
          "errors": Array [
            "Value should be a string",
          ],
          "path": "city",
        },
        Object {
          "children": null,
          "errors": Array [
            "Either Value should be a string OR Value should be a number",
            "Value should be a string",
            "Value should be a number",
          ],
          "path": "house",
        },
      ],
      "errors": Array [
        "Either Value should have a valid shape OR Value should be a string",
        "Value should have a valid shape",
        "Value should be a string",
      ],
      "path": "address",
    },
  ],
  "errors": Array [
    "Value should have a valid shape",
  ],
  "path": "",
}
`;

exports[`groupErrors() should work for single error 1`] = `
Object {
  "children": null,
  "errors": Array [
    "Value should be a string",
  ],
  "path": "",
}
`;

exports[`groupErrors() should work for single-path errors 1`] = `
Object {
  "children": null,
  "errors": Array [
    "Either Value should be a string OR Value should be a number",
    "Value should be a string",
    "Value should be a number",
  ],
  "path": "",
}
`;
